Mathlib/CategoryTheory/Galois/Basic.lean

```diff
@@ -0,0 +1,149 @@
+we omit the factorisation of morphisms in epimorphisms and monomorphisms
+as this is not needed for the proof of the fundamental theorem on Galois categories
+(and then follows from it).
+
+## References
+
+* H. W. Lenstra. Galois theory for schemes
+  <https://websites.math.leidenuniv.nl/algebra/GSchemes.pdf>
+
+-/
+
+universe u₁ u₂ v₁ v₂ w
+
+open CategoryTheory Limits Functor
+
+namespace Galois
+
+section Def
+
+/-- Definition of a Galois category. Lenstra, Def 3.1, (G1)-(G3) -/
+class PreGaloisCategory (C : Type u₁) [Category.{u₂, u₁} C] : Prop where
+  /-- `C` has a terminal object (G1). -/
+  hasTerminalObject : HasTerminal C
+  /-- `C` has pullbacks (G1). -/
+  hasPullbacks : HasPullbacks C
+  /-- `C` has finite coproducts (G2). -/
+  hasFiniteCoproducts : HasFiniteCoproducts C
+  /-- `C` has quotients by finite groups (G2). -/
+  hasQuotientsByFiniteGroups (G : Type u₂) [Group G] [Finite G] :
+    HasColimitsOfShape (SingleObj G) C
+  /-- Every monomorphism in `C` induces an isomorphism on a direct summand (G3). -/
+  monoInducesIsoOnDirectSummand {X Y : C} (i : X ⟶ Y) [Mono i] : ∃ (Z : C) (u : Z ⟶ Y),
+    Nonempty (IsColimit (BinaryCofan.mk i u))
+
+/-- Definition of a fibre functor from a Galois category. Lenstra, Def 3.1, (G4)-(G6) -/
+class FibreFunctor {C : Type u₁} [Category.{u₂, u₁} C] [PreGaloisCategory C]
+    (F : C ⥤ FintypeCat.{w}) where
+  /-- `F` preserves terminal objects (G4). -/
+  preservesTerminalObjects : PreservesLimitsOfShape (CategoryTheory.Discrete PEmpty.{1}) F
+  /-- `F` preserves pullbacks (G4). -/
+  preservesPullbacks : PreservesLimitsOfShape WalkingCospan F
+  /-- `F` preserves finite coproducts (G5). -/
+  preservesFiniteCoproducts : PreservesFiniteCoproducts F
+  /-- `F` preserves epimorphisms (G5). -/
+  preservesEpis : Functor.PreservesEpimorphisms F
+  /-- `F` preserves quotients by finite groups (G5). -/
+  preservesQuotientsByFiniteGroups (G : Type u₂) [Group G] [Finite G] :
+    PreservesColimitsOfShape (SingleObj G) F
+  /-- `F` reflects isomorphisms (G6). -/
+  reflectsIsos : ReflectsIsomorphisms F
+
+/-- An object of a category `C` is connected if it is not initial
+and has no non-trivial subobjects. Lenstra, 3.12. -/
+class ConnectedObject {C : Type u₁} [Category.{u₂, u₁} C] (X : C) : Prop where
+  /-- `X` is not an initial object. -/
+  notInitial : IsInitial X → False
+  /-- `X` has no non-trivial subobjects. -/
+  noTrivialComponent (Y : C) (i : Y ⟶ X) [Mono i] : (IsInitial Y → False) → IsIso i
+
+/-- A functor is said to preserve connected objects if it sends
+connected objects to connected objects. -/
+class PreservesConnectedObjects {C : Type u₁} [Category.{u₂, u₁} C] {D : Type v₁}
+    [Category.{v₂, v₁} D] (F : C ⥤ D) : Prop where
+  preserves : ∀ {X : C} [ConnectedObject X], ConnectedObject (F.obj X)
+
+end Def
+
+section Instances
+
+variable {C : Type u₁} [Category.{u₂, u₁} C] [PreGaloisCategory C]
+
+instance : HasTerminal C := PreGaloisCategory.hasTerminalObject
```

Comment:
I would suggest putting such instances in the `PregaloisCategory` namespace, and then it should be sufficient to add a line like `attribute [instance] hasTerminal hasPullbacks`, etc.

Reply:
Do you also suggest to drop the `namespace Galois` then?

Reply:
Yes. First, everything should be in the root `CategoryTheory` namespace. `PreGaloisCategory` should be just under `CategoryTheory`. Then, I would think most of the definitions in this file should be in `CategoryTheory.PreGaloisCategory`.

Reply:
I think I changed it accordingly.